consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java

```diff
@@ -14,11 +14,133 @@
+  private final BlockValidator blockValidator;
+  private final ProtocolContext protocolContext;
+
+  public RoundChangeMessageValidator(
+      final RoundChangePayloadValidator roundChangePayloadValidator,
+      final long quorumMessageCount,
+      final long chainHeight,
+      final Collection<Address> validators,
+      final BlockValidator blockValidator,
+      final ProtocolContext protocolContext) {
+    this.roundChangePayloadValidator = roundChangePayloadValidator;
+    this.quorumMessageCount = quorumMessageCount;
+    this.chainHeight = chainHeight;
+    this.validators = validators;
+    this.blockValidator = blockValidator;
+    this.protocolContext = protocolContext;
+  }
+
+  public boolean validate(final RoundChange msg) {
+
+    if (!roundChangePayloadValidator.validate(msg.getSignedPayload())) {
+      LOG.info("{}: embedded payload was invalid", ERROR_PREFIX);
+      return false;
+    }
+
+    if (msg.getProposedBlock().isPresent()) {
+      return validateWithBlock(msg);
+    }
+
+    return msg.getPreparedRoundMetadata().isEmpty();
+  }
+
+  private boolean validateBlock(final Block block) {
+    final Optional<BlockProcessingOutputs> validationResult =
+        blockValidator.validateAndProcessBlock(
+            protocolContext, block, HeaderValidationMode.LIGHT, HeaderValidationMode.FULL);
+
+    if (validationResult.isEmpty()) {
+      LOG.info("{}: block did not pass validation.", ERROR_PREFIX);
+      return false;
+    }
+
     return true;
   }
+
+  private boolean validateWithBlock(final RoundChange msg) {
+    final Block block = msg.getProposedBlock().get();
+
+    if (!validateBlock(block)) {
+      return false;
+    }
+
+    if (msg.getPreparedRoundMetadata().isEmpty()) {
+      LOG.info("{}: Prepared block specified, but prepared metadata absent", ERROR_PREFIX);
+      return false;
+    }
+
+    final PreparedRoundMetadata metadata = msg.getPreparedRoundMetadata().get();
+
+    if (!metadata.getPreparedBlockHash().equals(block.getHash())) {
+      LOG.info("{}: Prepared metadata hash does not match supplied block");
+      return false;
+    }
+
+    final BftExtraData extraData = BftExtraData.decode(block.getHeader());
+    if (metadata.getPreparedRound() != extraData.getRound()) {
+      LOG.info(
+          "{}: Prepared metadata contains a different round to that in the supplied block",
+          ERROR_PREFIX);
+      return false;
+    }
+
+    return validatePrepares(metadata, msg.getPrepares());
+  }
+
+  private boolean validatePrepares(
+      final PreparedRoundMetadata metaData, final List<SignedData<PreparePayload>> prepares) {
+
+    final ConsensusRoundIdentifier preparedRoundIdentifier =
+        new ConsensusRoundIdentifier(chainHeight, metaData.getPreparedRound());
+
+    final PrepareValidator validator =
+        new PrepareValidator(validators, preparedRoundIdentifier, metaData.getPreparedBlockHash());
+
+    if (hasDuplicateAuthors(prepares)) {
```

Comment:
I think duplicate messages are allowed, they just need to be counted as one when determining quorum.